qiskit_experiments/library/tomography/fitters/cvxpy_utils.py

```diff
@@ -51,54 +64,168 @@ def decorated_func(*args, **kwargs):
-            if HAS_CVXPY:
-                # pylint:disable=import-error
-                solvers = cvxpy.installed_solvers()
-                # Check for other SDP solvers cvxpy supports
-                for solver in ["CVXOPT", "MOSEK"]:
-                    if solver in solvers:
-                        cls._SDP_SOLVERS.add(solver)
-                        cls._HAS_SDP_SOLVER = True
-                        cls._HAS_SDP_SOLVER_NOT_SCS = True
-                if "SCS" in solvers:
-                    # Try example problem to see if built with BLAS
-                    # SCS solver cannot solver larger than 2x2 matrix
-                    # problems without BLAS
-                    try:
-                        var = cvxpy.Variable((5, 5), PSD=True)
-                        obj = cvxpy.Minimize(cvxpy.norm(var))
-                        cvxpy.Problem(obj).solve(solver="SCS")
-                        cls._SDP_SOLVERS.add("SCS")
-                        cls._HAS_SDP_SOLVER = True
-                    except cvxpy.error.SolverError:
-                        pass
+def solve_iteratively(problem: Problem, iters: int, **solve_kwargs) -> None:
+    """Solve a CVXPY problem increasing iterations if solution is inaccurate.
+
+    If the problem is not solved with the initial ``iters`` value of
+    iterations the number of iterations will be doubled up to the
+    specified ``"max_iters"`` value in the solve_kwargs. If no max
+    iters is specified this will be set to 4 times the initial iters
+    values.
+
+    Args:
+        problem: A CVXPY Problem to solve
+        iters: The initial number of max iterations to use when solving
+               the problem
+        solve_kwargs: kwargs for problem.solve method.
+
+    Raises:
+        AnalysisError: If the CVXPY solve fails to return an optimal or
+                       optimal_inaccurate solution.
+    """
+    max_iters = solve_kwargs.get("max_iters", 4 * iters)
+    problem_solved = False
+    while not problem_solved:
+        solve_kwargs["max_iters"] = iters
+        problem.solve(**solve_kwargs)
+        if problem.status in ["optimal_inaccurate", "optimal"]:
+            problem_solved = True
+        elif problem.status == "unbounded_inaccurate":
+            if iters < max_iters:
+                iters *= 2
+            else:
+                raise AnalysisError(
+                    "CVXPY fit failed, probably not enough iterations for the solver"
+                )
+        elif problem.status in ["infeasible", "unbounded"]:
+            raise AnalysisError(
+                "CVXPY fit failed, problem status {} which should not happen".format(problem.status)
+            )
+        else:
+            raise AnalysisError("CVXPY fit failed, reason unknown")
+
+
+def set_default_sdp_solver(solver_kwargs: dict):
+    """Set default SDP solver from installed solvers."""
+    if "solver" in solver_kwargs:
+        return
+    if "CVXOPT" in cvxpy.installed_solvers():
```

Comment:
This is not analysis option with default value? This branching seems bit weird since we do nothing when CVXOPT and MOSEK are not available. This is intended to let cvxopt use its preferred solver?

Reply:
Its a cvxpy solver option (which go into the fitter_options dict in analysis options), which depends on an optional library being installed.

The point of this is that if no solver is specified the default SDP solver of CVXPY is SCS, which isn't very good compared to the optional ones that must be installed separately because of their licenses. So if a better one is already installed and available we override the default to use it by passing it in as a solver kwarg.

Reply:
Could you write this down as a comment or function docstring? I misunderstood the purpose of this function. Probably you can raise user warning SCS may give you poor result, or probably you can copy this solver to analysis metadata, i.e. extra.

Reply:
The cvxpy solver is in the analysis result extra field under "cvxpy_solver".

SCS may be ok so I don't think it warrants a warning, but it is a first order method, when the cvxopt solver is a 2nd order method so has longer runtime but more accuracy, but I think that makes it a better default if it's available, thus the override.